### PR TITLE
MINOR: Update dependencies

### PR DESCRIPTION
**Pull request by jlprat**
Updates minor versions for our dependencies and build tool

- Jackson from 2.16.0 to 2.16.1 https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.16.1
- JUnit from 5.10.0 to 5.10.2 https://junit.org/junit5/docs/5.10.2/release-notes/ and https://junit.org/junit5/docs/5.10.1/release-notes/
- Mockito from 5.8.0 to 5.10.0 (only if JDK 11 or higher) https://github.com/mockito/mockito/releases/tag/v5.10.0 and https://github.com/mockito/mockito/releases/tag/v5.9.0
- Gradle from 8.5 to 8.6 https://docs.gradle.org/8.6/release-notes.html

Gradle upgrade was done via:
`./gradlew wrapper --gradle-version latest`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: gradlew
Comment (jlprat) [inline gradlew:1]: Changes in this file are auto-generated

**Comment by divijvaidya** [comment]
Please update license of jackson here - https://github.com/apache/kafka/blob/cc49fc76564b15f2111801faf62341aaf01c4d8c/LICENSE-binary#L219

File: gradlew [hunk 1]
```diff
@@ -215,11 +215,11 @@ fi
 # Add default JVM options here. You can also use JAVA_OPTS and GRADLE_OPTS to pass JVM options to this script.
 DEFAULT_JVM_OPTS='"-Xmx64m" "-Xms64m"'
 
-# Collect all arguments for the java command;
-#   * $DEFAULT_JVM_OPTS, $JAVA_OPTS, and $GRADLE_OPTS can contain fragments of
-#     shell script including quotes and variable substitutions, so put them in
-#     double quotes to make sure that they get re-expanded; and
-#   * put everything else in single quotes, so that it's not re-expanded.
+# Collect all arguments for the java command:
```
Comment (divijvaidya) [inline gradlew:218]: A similar change was made in another PR and turns out that these lines may not need to be changed. Here is a reference to @ijuma's comment about it https://github.com/apache/kafka/pull/14502#discussion_r1355560355 

Can you please add what commands are you running that is leading to change in these lines?

File: gradle/dependencies.gradle [hunk 2]
```diff
@@ -103,7 +103,7 @@ versions += [
   gradle: "8.5",
   grgit: "4.1.1",
   httpclient: "4.5.14",
-  jackson: "2.16.0",
+  jackson: "2.16.1",
```
Comment (divijvaidya) [inline gradle/dependencies.gradle:106]: please add jackson diff to the description of the PR for completeness - https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.16.1 

I have reviewed the diff and I don't see any problems in this patch version.

File: gradlew (hunk 1 above)
Comment (jlprat) [reply gradlew:218]: This is what I run to upgrade to the latest version of Gradle (as per documentation): https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:upgrading_wrapper
`./gradlew wrapper --gradle-version latest`

**Comment by jlprat** [comment]
I would say `trunk` build is broken again. I'll investigate

**Comment by divijvaidya** [comment]
It must have been fixed by https://github.com/apache/kafka/commit/cc49fc76564b15f2111801faf62341aaf01c4d8c 
Perhaps a rebase from trunk should help.

**Comment by jlprat** [comment]
Rebased

**Comment by jlprat** [comment]
@divijvaidya as you already partially reviewed this, do you mind finishing the review?
Also to which versions did you think this should be backported?

**Review by divijvaidya — APPROVED** [review]
Thanks for the change @jlprat. 

Regarding backporting, the motivation is to keep our bugfix versions free on any CVEs caused by a dependency, hence, we prefer to backport minor upgrades to previous  versions. In this case, backporting to 3.7.x and 3.6.x should be sufficient, since it's unlikely we are going to release 3.5 bug fix.

**Comment by jlprat** [comment]
Backported to `3.7` branch

**Comment by jlprat** [comment]
Regarding a backport to the `3.6` branch, Gradle version over there is `8.2.1`, it's a bit of a jump I believe.

**Comment by divijvaidya** [comment]
> Regarding a backport to the 3.6 branch, Gradle version over there is 8.2.1, it's a bit of a jump I believe.

Ok, no worries. Let's keep it limited to 3.7.x. We can always revisit updating dependencies if/when we start a patch fix for 3.6.2

File: gradle/dependencies.gradle [hunk 3]
```diff
@@ -100,10 +100,10 @@ versions += [
   commonsCli: "1.4",
   commonsValidator: "1.7",
   dropwizardMetrics: "4.1.12.1",
-  gradle: "8.5",
+  gradle: "8.6",
```
Comment (pasharik) [inline gradle/dependencies.gradle:103]: This PR has already been merged, but I'll leave a comment here ahyway.
I've noticed, what behaviour of incremental compilation for scala tests in gradle seems to be changed after gradle version upgrade from `8.5` to `8.6`.

With gradle `8.6`, if I do any change to some unit-test, e.g. `kafka.admin.AclCommandTest.scala`, and then run:

```
./gradlew core:compileTestScala
```

the entire folder `core/build/classes/scala/test` is cleared, and gradle re-compiles _all_ tests. It takes quite a long time. So the incremental compilation for scala tests seems to be broken.

If I change Gradle version back to `8.5`, incremental compilation works as expected

Comment (ijuma) [reply gradle/dependencies.gradle:103]: That's a big deal, we should probably revert to the older version.

Comment (jlprat) [reply gradle/dependencies.gradle:103]: https://github.com/apache/kafka/pull/15553 PR reverting to the old version.

@pasharik have you reported the bug (if it is indeed a bug in Gradle) to the upstream project?

Comment (jlprat) [reply gradle/dependencies.gradle:103]: I see some inconsistencies with this report.
I managed to reproduce it only once. But when I tried it again, incremental compilation worked as expected.

Comment (pasharik) [reply gradle/dependencies.gradle:103]: @jlprat my `trunk` branch was almost 2 weeks old, now I've updated it and I can't see any difference between Gradle 8.5 and 8.6. So I think we can leave this PR as it is, sorry for the confusion.

I didn't report issue to the Gradle yet, still trying to figure out exact steps how to reproduce it, and if it's related to Gradle or to my local environment setup.

Not related to this particular PR, but I still can see issues with incremental compilation, only when making changes from IntelliJ Idea. If changes are made from vim or other console text editor, incremental compilation works as expected.

Steps to reproduce:

* Open `kafka.admin.AclCommandTest` in IntelliJ Idea. I'm using `2023.3.5 Ultimate`
* Add `println("a")` at the beginning of `testAclCliWithAuthorizer` method
* Run in terminal: `./gradlew core:test --tests "kafka.admin.AclCommandTest.testAclCliWithAuthorizer"`
* Result: all tests in `core` module are re-compiled

If test modified with `vim`, only single test class is re-compiled

**Comment by pasharik** [comment]
There is a version of gradle wrapper in our [gradlew](https://github.com/apache/kafka/blob/trunk/gradlew#L119) file:

```
https://raw.githubusercontent.com/gradle/gradle/v8.5.0/gradle/wrapper/gradle-wrapper.jar
```

should we update to `8.6.0` as well?

**Comment by jlprat** [comment]
I generated the wrapper the way `gradle` recommends (`./gradlew wrapper --gradle-version latest`). I would leave it the way gradle generates it

**Pull request merged**